### ISSUE TITLE
Rule proposal: `no-this-outside-of-class`

### ISSUE THREAD

**Issue by zanminkian** [sindresorhus/eslint-plugin-unicorn#2450]
### Description

In modern JavaScript, `this` should not appear outside of an class.

### Fail

```js
function foo () {
  this.bar;
}
function Foo(name) {
  this.name = name;
}
class Foo {
  foo() {
    function bar() {
      this.baz;
    }
  }
}
```

### Pass

```js
class Foo {
  foo() {
    this.name;
  }
}

class Foo {
  foo() {
    const bar = () => {
      this.name;
    }
  }
}
```

### Proposed rule name

no-this-outside-of-class

**Comment by axetroy** [comment]
This rule should be easy to implement. It is similar to this rule #2525

Once this rule is accepted, I will work on it.

**Comment by axetroy** [comment]
@sindresorhus Is this rule acceptable?

**Comment by sindresorhus** [comment]
Accepted

**Comment by fisker** [comment]
Should we exclude class modification pattern?

```js
Foo.prototype.bar = function () {
	this.baz()
}
``

**Comment by fisker** [comment]
> I don't think modifying prototype manually in morden JavaScript is a good practice.

What's the alternative? Subclass shouldn't the same thing.

**Comment by fisker** [comment]
Also, should we allow `this` in object methods?

**Comment by axetroy** [comment]
Should we allow `this` in the top-level scope?

Many umd modules and some special environments may not have `window` and `global,` and this is needed for detection

**Comment by axetroy** [comment]
> Also, should we allow `this` in object methods?

It should be allowed, Vue2 options is this mode, otherwise there are many false positives

```vue
<script>
export default {
  methods: {
    foo () {
      this.name = 'xxx'
    }
  }
}
</script>
```

**Comment by axetroy** [comment]
Examples of false positives

```js
new SDK({
    appKey: 'xxx',
    appId: 'xxx',
    onReady: function () {
        this.xxxxx
    }
})
```

```vue


<script>
export default {
  methods: {
    refresh: debounce(async function () {
       this.list = await getList()
    })
  }
}
</script>
```

**Comment by axetroy** [comment]
Vue2 is EOL, But Vue's options mode is not.

And similar situations

```js
export defaults defineComponent({
  methods: {
    mounted() {
        this.xxxxxx // 
    }
  }
})
```

Maybe we should wait for more feedback.
